docs/man/man3/uuid.3
.TH "uuid" 3 "Wed Jan 25 2023" "Version 00.01a07-dbg" "Freecraft" \" -*- nroff -*-
.ad l
.nh
.SH NAME
uuid
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <uuid\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBsetrandomly\fP (void)"
.br
.RI "set uuid randomly\&. "
.ti -1c
.RI "\fBuuid\fP (void)"
.br
.RI "Construct a new uuid object\&. "
.ti -1c
.RI "\fBuuid\fP (const \fBuuid\fP &dat)"
.br
.RI "Construct a new uuid object\&. "
.ti -1c
.RI "\fBuuid\fP (const Json::String &str)"
.br
.RI "Construct a new uuid object\&. "
.ti -1c
.RI "const \fBuuid\fP & \fBoperator=\fP (const \fBuuid\fP &dat)"
.br
.RI "assignment operator overloading\&. "
.ti -1c
.RI "bool \fBoperator==\fP (const \fBuuid\fP &dat) const"
.br
.RI "check if two uuids is equal\&. "
.ti -1c
.RI "\fBoperator Json::String\fP (void) const"
.br
.RI "conversation function\&. "
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "char \fBdata\fP [37]"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "bool \fBoperator==\fP (const Json::String &datL, const \fBuuid\fP &datR)"
.br
.RI "same\&. "
.in -1c
.SH "Detailed Description"
.PP 
Definition at line \fB11\fP of file \fBuuid\&.hpp\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "uuid::uuid (void)"

.PP
Construct a new uuid object\&. 
.PP
Definition at line \fB60\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
61 {
62     for(int i=0; i < 37; ++i)
63         data[i]=0;
64 }
.fi

.SS "uuid::uuid (const \fBuuid\fP & dat)"

.PP
Construct a new uuid object\&. 
.PP
\fBParameters\fP
.RS 4
\fIdat\fP 
.RE
.PP

.PP
Definition at line \fB66\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
67 {
68     for(int i=0; i < 37; ++i)
69         data[i]=dat\&.data[i];
70 }
.fi

.SS "uuid::uuid (const Json::String & str)"

.PP
Construct a new uuid object\&. 
.PP
\fBParameters\fP
.RS 4
\fIstr\fP 
.RE
.PP

.PP
Definition at line \fB99\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
100 {
101     if(!checksyntax(str))
102         throw fcSyntaxException("UUID format is not wanted");
103     for(int i=0; i < str\&.length(); ++i)
104         data[i]=str[i];
105     data[str\&.length()]='\\0';
106 }
.fi

.SH "Member Function Documentation"
.PP 
.SS "uuid::operator Json::String (void) const"

.PP
conversation function\&. 
.PP
Definition at line \fB108\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
109 {
110     return Json::String(data);
111 }
.fi

.SS "const \fBuuid\fP & uuid::operator= (const \fBuuid\fP & dat)"

.PP
assignment operator overloading\&. 
.PP
\fBParameters\fP
.RS 4
\fIdat\fP uuid to assign\&. 
.RE
.PP
\fBReturns\fP
.RS 4
const uuid& 
.RE
.PP

.PP
Definition at line \fB123\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
124 {
125     for(int i=0; i<37; ++i)
126         data[i] = dat\&.data[i];
127 }
.fi

.SS "bool uuid::operator== (const \fBuuid\fP & dat) const"

.PP
check if two uuids is equal\&. 
.PP
\fBParameters\fP
.RS 4
\fIdat\fP 
.RE
.PP
\fBReturns\fP
.RS 4
true equal\&. 
.PP
false not equal\&. 
.RE
.PP

.PP
Definition at line \fB113\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
114 {
115     return dat\&.data == data;
116 }
.fi

.SS "void uuid::setrandomly (void)"

.PP
set uuid randomly\&. 
.PP
Definition at line \fB39\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
40 {
41     std::default_random_engine e;int i=0;
42     std::uniform_int_distribution<unsigned int> u(0,15);
43     for(i=0; i <= 7; ++i)
44         data[i]=uitchex(u(e));
45     data[8]='\-';
46     for(i=9; i <= 12; ++i)
47         data[i]=uitchex(u(e));
48     data[13]='\-';
49     for(i=14; i <= 17; ++i)
50         data[i]=uitchex(u(e));
51     data[18]='\-';
52     for(i=19; i <= 22; ++i)
53         data[i]=uitchex(u(e));
54     data[23]='\-';
55     for(i=24; i <= 35; ++i)
56         data[i]=uitchex(u(e));
57     data[36]='\\0';
58 }
.fi

.SH "Friends And Related Function Documentation"
.PP 
.SS "bool operator== (const Json::String & datL, const \fBuuid\fP & datR)\fC [friend]\fP"

.PP
same\&. reversed\&.
.PP
\fBParameters\fP
.RS 4
\fIdat\fP 
.br
\fIdat\fP 
.RE
.PP
\fBReturns\fP
.RS 4
true 
.PP
false 
.RE
.PP

.PP
Definition at line \fB118\fP of file \fBuuid\&.cpp\fP\&..PP
.nf
119 {
120     return datR\&.data == datL;
121 }
.fi

.SH "Member Data Documentation"
.PP 
.SS "char uuid::data[37]\fC [protected]\fP"

.PP
Definition at line \fB14\fP of file \fBuuid\&.hpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Freecraft from the source code\&.


src/core/data/uuid/uuid.cpp
/**
 * @file uuid.cpp
 * @author Promesis (m3149154722@outlook.com)
 * @brief 
 * @version 00.01b01-dev
 * @date 2023-01-27
 * 
 * Promesis 2023
 * the Freecraft dev
 * 
 */

#include "uuid.hpp"
//xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx

//moved from 00.01a07-dbg
inline char uitchex(unsigned int i)
{
    switch(i)
    {
        case 0: return '0'; break;
        case 1: return '1'; break;
        case 2: return '2'; break;
        case 3: return '3'; break;
        case 4: return '4'; break;
        case 5: return '5'; break;
        case 6: return '6'; break;
        case 7: return '7'; break;
        case 8: return '8'; break;
        case 9: return '9'; break;
        case 10: return 'a'; break;
        case 11: return 'b'; break;
        case 12: return 'c'; break;
        case 13: return 'd'; break;
        case 14: return 'e'; break;
        case 15: return 'f'; break;
        default: return '\0'; break;
    }
}


inline bool isHexChar(char c)
{
    char ch=std::isalpha(c)? std::tolower(c) : c;
    switch(ch)
    {
        case 'a':   case 'b':   case 'c':   case 'd':   case 'e':   case 'f':
        case '0':   case '1':   case '2':   case '3':   case '4':   case '5':
        case '6':   case '7':   case '8':   case '9': 
        {
            return true;
            break;
        }
        default:
        {
            return false;
            break;
        }
    }
}

//cast
inline bool ishexnum(char c)
{
    return isHexChar(c);
}

//moved from 00.01a07-dbg
inline bool checksyntax(const std::string &str)
{
    if(str.length() != 36)
        return false;
    for(i=0; i <= 7; ++i)
        if(!ishexnum(str[i]))
            return false;
    if(str[8]!='-') return false;
    for(i=9; i <= 12; ++i)
        if(!ishexnum(str[i]))
            return false;
    if(str[13]!='-') return false;
    for(i=14; i <= 17; ++i)
        if(!ishexnum(str[i]))
            return false;
    if(str[18]!='-') return false;
    for(i=19; i <= 22; ++i)
        if(!ishexnum(str[i]))
            return false;
    if(str[23]!='-') return false;
    for(i=24; i <= 35; ++i)
        if(!ishexnum(str[i]))
            return false;
    return true;
}

fcUUID::fcUUID(void)
{
    data_="00000000000000000000000000000000";
}

fcUUID::fcUUID(const std::string &data)
{
    if(!checksyntax(data))
    {
        ferr << "UUID grammar isn't right" << fcendl;
        return;
    }
    for(int i = 0; i < 8; ++i)
        data_[i] = data[i];
    for(int i = 8; i < 12; ++i)
        data_[i] = data[i+1];
    for(int i = 12; i < 16; ++i)
        data_[i] = data[i+2];
    for(int i = 16; i < 20; ++i)
        data_[i] = data[i+3];
    for(int i = 20; i < 24; ++i)
        data_[i] = data[i+4];
    for(int i = 24; i < 28; ++i)
        data_[i] = data[i+5];
    for(int i = 28; i < 32; ++i)
        data_[i] = data[i+6];
}

const fcUUID & newUUID( void )
{
    std::default_random_engine eng;
    std::uniform_int_distribution< unsigned int > uni(0,15);
    fcUUID retVal;
    for(int i = 0; i < 32; ++i)
    {
        retVal.data_[i] = uitchex( uni( eng ) );
    }
}